src/query/api/v1/handler/placement/set.go

```diff
@@ -0,0 +1,150 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package placement
+
+import (
+	"net/http"
+	"path"
+	"time"
+
+	"github.com/m3db/m3/src/cluster/placement"
+
+	"github.com/m3db/m3/src/query/api/v1/handler"
+	"github.com/m3db/m3/src/query/api/v1/handler/prometheus/handleroptions"
+	"github.com/m3db/m3/src/query/generated/proto/admin"
+	"github.com/m3db/m3/src/query/util/logging"
+	xhttp "github.com/m3db/m3/src/x/net/http"
+
+	"github.com/gogo/protobuf/jsonpb"
+	"go.uber.org/zap"
+)
+
+const (
+	// SetHTTPMethod is the HTTP method for the the upsert endpoint.
+	SetHTTPMethod = http.MethodPost
+
+	setPathName = "set"
+)
+
+var (
+	// M3DBSetURL is the url for the m3db replace handler (method POST).
+	M3DBSetURL = path.Join(handler.RoutePrefixV1,
+		M3DBServicePlacementPathName, setPathName)
+
+	// M3AggSetURL is the url for the m3aggregator replace handler (method
+	// POST).
+	M3AggSetURL = path.Join(handler.RoutePrefixV1,
+		handleroptions.M3AggregatorServiceName, setPathName)
+
+	// M3CoordinatorSetURL is the url for the m3coordinator replace handler
+	// (method POST).
+	M3CoordinatorSetURL = path.Join(handler.RoutePrefixV1,
+		handleroptions.M3CoordinatorServiceName, setPathName)
+)
+
+// SetHandler is the type for placement replaces.
+type SetHandler Handler
+
+// NewSetHandler returns a new SetHandler.
+func NewSetHandler(opts HandlerOptions) *SetHandler {
+	return &SetHandler{HandlerOptions: opts, nowFn: time.Now}
+}
+
+func (h *SetHandler) ServeHTTP(
+	svc handleroptions.ServiceNameAndDefaults,
+	w http.ResponseWriter,
+	r *http.Request,
+) {
+	ctx := r.Context()
+	logger := logging.WithContext(ctx, h.instrumentOptions)
+
+	req, pErr := h.parseRequest(r)
+	if pErr != nil {
+		xhttp.Error(w, pErr.Inner(), pErr.Code())
+		return
+	}
+
+	serviceOpts := handleroptions.NewServiceOptions(svc,
+		r.Header, h.m3AggServiceOptions)
+	service, _, err := ServiceWithAlgo(h.clusterClient,
+		serviceOpts, h.nowFn(), nil)
+	if err != nil {
+		logger.Error("unable to create placement service", zap.Error(err))
+		xhttp.Error(w, err, http.StatusInternalServerError)
+		return
+	}
+
+	curPlacement, err := service.Placement()
+	if err != nil {
+		logger.Error("unable to get current placement", zap.Error(err))
+		xhttp.Error(w, err, http.StatusInternalServerError)
```

Comment:
Let's make this `StatusNotFound`? Thinking from the operator perspective, there are times where we have to type check a 404 from Coordinator API.

Reply:
Sounds good, I added this.